mission_control/src/behaviors/abort_behavior.cpp

```diff
@@ -0,0 +1,113 @@
+/*********************************************************************
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2021, QinetiQ, Inc.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of QinetiQ nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+// Original version: Christopher Scianna Christopher.Scianna@us.QinetiQ.com
+#include <string>
+
+#include "mission_control/behaviors/abort_behavior.h"
+
+using mission_control::AbortBehavior;
+
+AbortBehavior::AbortBehavior(const std::string& name,
+                                             const BT::NodeConfiguration& config)
+    : Behavior(name, config)
+{
+  subStateData_ =
+      nodeHandle_.subscribe("/state", 1, &AbortBehavior::stateDataCallback, this);
+
+  subThrusterRPM_ = nodeHandle_.subscribe("/thruster_control/report_rpm", 1,
+                                          &AbortBehavior::thrusterRPMCallback, this);
+
+  attitudeServoBehaviorPub_ =
+      nodeHandle_.advertise<mission_control::AttitudeServo>("/mngr/attitude_servo", 1);
+
+}
+
+BT::NodeStatus AbortBehavior::behaviorRunningProcess()
+{
+  if (!abortMsgHasBeenPublished_)
```

Comment:
@patapufate meta: this pattern of using an internal state variable to do "initial" work pops up everywhere. We should probably be using the IDLE -> RUNNING transition to do this (which `Behavior` uses to switch state and do nothing). I'll open a PR with waypoint following support soon that attempts this slight shift.

Reply:
Changed